### DISCUSSION TITLE
Schema.org Compatibility - LaTeX Export offering

### DISCUSSION THREAD

**Discussion by supaeasy** [TandoorRecipes/recipes#1462] [Ideas]
Hello there

I just stumbled across this marvellous project (thank you [heise.de](https://www.heise.de/ratgeber/Open-Source-Software-Tandoor-Recipes-Rezeptdatenbank-einfach-selbst-hosten-6319271.html)) and it looks like everything I ever wanted in a recipe database. 

While I love the Idea of having a flexible database with scaling, shopping list and whatnot I also heavily rely on printed recipes in a decent look. Right now I am using a fiddly, fiddly workflow of storing recipes in Paprika App, exporting them to json and parsing them in Airtable to LaTeX via Overleaf. While this does get me the beatuiful end result I want (see example attached), this is nowhere near convenient and I see this project as a real opportunity for a single-step-export solution. 

I will be happy to share my parsing method from json to LaTeX but am no real programmer myself and don't know how to implement this as an export function to Tandoor Recipes. Is this something you could do? It would however require TR to comply with schema.org's [Recipe Standard](https://schema.org/Recipe) for a quick start. If it does not comply, I would still offer to rewrite my method to TR's system if an export to LaTeX function would be considered.

[Kochbuch_xcookybooky (1).pdf](https://github.com/TandoorRecipes/recipes/files/7978561/Kochbuch_xcookybooky.1.pdf)

**Comment by StephenBrown2** [comment]
Have you taken a look at the underlying schemas (https://github.com/TandoorRecipes/recipes/blob/develop/cookbook/models.py#L473 and below) or some of the export functions (for example https://github.com/TandoorRecipes/recipes/blob/develop/cookbook/integration/recipesage.py#L61 looks to use the Schema.org Recipe schema when exporting)?

I believe when this project started it was basically just dumping the schema.org schema into a mongodb document, but the move to postgres as a backend did change some things. Still perfectly possible to get a standard schema out though, just takes putting the pieces together.

**Reply by supaeasy** [reply]
Thank you for pointing this out, especially the latter. Looks good to me an when comparing an export of Paprika and the export parameters here I must say that TR defnitely follows schema even closer. 

One thing I find odd is the definition of total time as waiting time + working time, since there are definitions for that: totalTime, prepTime, cookTime. One might argue that cooldown time or day-before-preparations for e.g. dough or cale being omitted here and thus shouldn't be calculated at all and instead these should be 3 hard values or am I missing out on sth?

In the end: cookTime is the only Variable I am missing. I will have to rewrite my stuff (very bad naming on Paprika side, I only see this now) but it should work just fine.

I am abusing the comment property to define Cooking utensils needed for the recipe since to my surprise there is no property defined for that.

**Comment by vabene1111** [comment]
first of all thanks for the nice words and the suggestion.

The time thingy comes from a time probably 3-4 years ago where i had no clue about schema.org recipe format and was just getting started with this application and just tought which times would be nice to track 😂 but i guess they can be mapped to schema org (working = prep and waiting=cook)

i think if you provide a latex template with the places where values would need to be inserted then it should be rather easy to create a working latex export. I myself am quite familiar with it, would just save me some work if you already have a working template. 

if you want to help get this done i would recommend you start by looking at the model (as @StephenBrown2 recommended) or at the respoonse of the `/api/recipe/<id>` json view and try to make a template that maps that "structure" to latex code. (mainly some kind of "main body" and something to support steps and ingredients). The rest is something i can quickly do afterwards. If there is a small and efficent enough latex compiler we might even be able to implement converting the results to pdf as well.

**Reply by supaeasy** [reply]
First of all: thank you for your quick answer, I appreciate it. Also: sorry for my late reply. I am struggling with moving right now and can only do so much in my free time. 

I have revisited my "code" and have to say it really is a mess. I will try to clean it up a little, but originally I thought of it as a temporary WIP solution - and then it never surpassed that stage. As I said I did it in Airtable it now really is a chain of roughly 25 formula fields each with successive subtitutions. It does work, however I will have to re-figure out, how. The easiest way would be to attach Text files of every formula field, but I doubt you will have the time and sense to dig through it yourself - or do you? First I will have to find out how many of these fields are actually used in the end formula. There are a lot of "copy of..." backups and I really need to wrap my head around it first.

**Reply by supaeasy** [reply]
Actually dicovering used fields went quicker than expected. Have a look at it (and despair, I guess)

All steps but two ("Schriftgrösse nach Länge" and "Ausgabe in TeX") reference to the JSON File and their results are processed later.

I think you speak german so you should get the Names as almost self-explaining, but I will try and document everything in english:

It is worth mentioning that I have made some custom "Spelling rules" for myself: 
It is necessary to write Steps in the recipe as "-Schritt 1-", "-Schritt 2-" and so on. Otherwise it will look ugly. "°C" and "100g"/l/ml has to be written like that, no space in between, Also no Abbreviations like "EL", but "Esslöffel" German plural forms like "(er)", "(en)", "(e)" for scaling, will be handled as well but they are not necessarily needed. In the field where nutritional info is supposed to be, i put needed Cooking utensils like "Large Pot, Pan". I hope I didnt't forget anything. (in everything: omit ""). 

- **JSON** is the raw JSON Export from Paprika - I have included an example
- **Schriftgrösse nach Länge** simply counts characters and determines if the end result will be set in 10 or 11pt.
- **Zubereitung** converts Steps into xcookybooky steps.  
- **Zutaten mit Müll** is probably the largest one: converts any units (metric) into xcookybooky units including nice fractions (like ⅛).
- **Titel** simply prints Recipe Title
- **Vorbereitungszeit** prints Preparation time
- **Kochzeit** prints cooking time
- **Portionen** prints servings
- **Quelle:** Determines if there is a source. If yes, it will print Quelle: (this is needed to determine whether a source pictogramm should be displayed or not in the output) 
- **Quelle URL** prints the URL of the recipe's source
- **Quelle** will print Source name of the recipe
- **Beilage** does not necessarily do what the german name says :-) it prints the comment in the Box above the recipe if there is one
- **Geschirr** prints needed kitchen utensils
- **Ausgabe in TeX** will print the final LaTeX-Code, referencing to aforementioned fields (This is the most important file, generating the actual LaTeX Source Code. You might just want to start from here)
- **LaTeX end Result** is not part of the script, just the example outcome from **Ausgabe in TeX**.

I hope this helps but I must admit it is a TOTAL mess.

Edited for better explaining and readability

[Vorbereitungszeit.txt](https://github.com/TandoorRecipes/recipes/files/8061438/Vorbereitungszeit.txt)
[Titel.txt](https://github.com/TandoorRecipes/recipes/files/8061439/Titel.txt)
[Ausgabe in TeX.txt](https://github.com/TandoorRecipes/recipes/files/8061440/Ausgabe.in.TeX.txt)
[Zutaten mit Müll.txt](https://github.com/TandoorRecipes/recipes/files/8061441/Zutaten.mit.Mull.txt)
[Schriftgrösse nach Länge.txt](https://github.com/TandoorRecipes/recipes/files/8061442/Schriftgrosse.nach.Lange.txt)
[Zubereitung.txt](https://github.com/TandoorRecipes/recipes/files/8061443/Zubereitung.txt)
[LaTeX end Result.txt](https://github.com/TandoorRecipes/recipes/files/8061444/LaTeX.end.Result.txt)
[Json.txt](https://github.com/TandoorRecipes/recipes/files/8061445/Json.txt)
[Geschirr.txt](https://github.com/TandoorRecipes/recipes/files/8061446/Geschirr.txt)
[Beilage.txt](https://github.com/TandoorRecipes/recipes/files/8061447/Beilage.txt)
[Quelle.txt](https://github.com/TandoorRecipes/recipes/files/8061448/Quelle.txt)
[Quelle_(Underscore means colon) .txt](https://github.com/TandoorRecipes/recipes/files/8061449/Quelle_.Underscore.means.colon.txt)
[Quelle URL.txt](https://github.com/TandoorRecipes/recipes/files/8061450/Quelle.URL.txt)
[Portionen.txt](https://github.com/TandoorRecipes/recipes/files/8061451/Portionen.txt)
[Kochzeit.txt](https://github.com/TandoorRecipes/recipes/files/8061452/Kochzeit.txt)
